### PR TITLE
Fix gas giant landing

### PR DESCRIPTION
**Pull request by TomGoodIdea**
**Bugfix:** This PR addresses issue #5663

## Fix Details
Picked up #5677 after its abandonment by the original author; unified new code for planet choosing with an existing function (https://github.com/endless-sky/endless-sky/pull/5677#discussion_r567486402).

## Testing Done
Landed on a gas giant in Remnant space. My escort without gaslining landed on another planet, so I wasn't able to change my flagship.

### CODE REVIEW

File: source/AI.cpp [hunk 1]
```diff
@@ -1023,6 +999,36 @@ int64_t AI::EnemyStrength(const Government *government)
...
+	if(system)
+	{
+		// Determine which, if any, planet with fuel or without fuel is closest.
+		double closest = numeric_limits<double>::infinity();
+		const Point &p = ship.Position();
+		for(const StellarObject &object : system->Objects())
+		{
+			const Planet *planet = object.GetPlanet();
+			if(object.HasSprite() && object.HasValidPlanet() && !planet->IsWormhole()
+					&& planet->CanLand(ship) && !(planet->HasFuelFor(ship) ^ refuel))
```
Comment (quyykk) [inline source/AI.cpp:1016]: A negated xor? interesting but plz no.

**Comment by TomGoodIdea** [comment]
Fixed an exploit at shop panels.

File: source/ShopPanel.cpp [hunk 2]
```diff
@@ -55,9 +55,9 @@ namespace {
 	constexpr int ICON_COLS = 4;
 	constexpr float ICON_SIZE = ICON_TILE - 8;
 
-	bool CanShowInSidebar(const Ship &ship, const System *here)
+	bool CanShowInSidebar(const Ship &ship, const Planet *here)
 	{
-		return ship.GetSystem() == here && !ship.IsDisabled();
+		return ship.GetPlanet() == here;
```
[the remark this answers is not in the document]
Comment (TomGoodIdea) [reply source/ShopPanel.cpp:60]: Can a landed ship be disabled?

[the remark this answers is not in the document]
Comment (quyykk) [reply source/ShopPanel.cpp:61]: Interesting: https://github.com/endless-sky/endless-sky/commit/b4dbec490a7ce03a60558dcb4f4e01ed8c08f466

[the remark this answers is not in the document]
Comment (quyykk) [reply source/ShopPanel.cpp:61]: (so yeah please do keep it)

**Comment by quyykk** [comment]
I'd say that the entry of a ship that isn't on the current planet should be darker, just like for ships not in the current system.

image

File: source/ShopPanel.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (TomGoodIdea) [reply source/ShopPanel.cpp:61]: Removing the check doesn't add disabled ships to the sidebar, I've just tested that.

File: source/PlayerInfo.cpp [hunk 3]
```diff
@@ -1431,10 +1431,6 @@ void PlayerInfo::Land(UI *ui)
 			++it;
 	}
 
-	// "Unload" all fighters, so they will get recharged, etc.
-	for(const shared_ptr<Ship> &ship : ships)
-		ship->UnloadBays();
```
Comment (quyykk) [inline source/PlayerInfo.cpp:1436]: By moving this in the loop below, carriers that can't land on the current planet won't get their fighters recharged even if the carrier landed on another planet. So I would suggest keeping it like it was, unless I'm missing something.

File: source/PlayerInfo.cpp [hunk 4]
```diff
@@ -1796,7 +1817,7 @@ void PlayerInfo::UpdateCargoCapacities()
 	int bunks = 0;
 	flagship = FlagshipPtr();
 	for(const shared_ptr<Ship> &ship : ships)
-		if(ship->GetSystem() == system && !ship->IsParked() && !ship->IsDisabled())
+		if(ship->GetPlanet() == planet && !ship->IsParked() && !ship->IsDisabled())
```
Comment (TomGoodIdea) [inline source/PlayerInfo.cpp:1820]: I guess we can remove this check:
```suggestion
		if(ship->GetPlanet() == planet && !ship->IsParked())
```

Comment (quyykk) [reply source/PlayerInfo.cpp:1820]: seems like it

**Review by quyykk — APPROVED** [review]
thanks!

**Pull request merged**